pkg/controller/admissionchecks/multikueue/workload.go

```diff
@@ -279,11 +293,43 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
 			Reason:  remoteFinishedCond.Reason,
 			Message: fmt.Sprintf("From remote %q: %s", remote, remoteFinishedCond.Message),
 		})
-		return a.client.Status().Patch(ctx, wlPatch, client.Apply, client.FieldOwner(ControllerName+"-finish"), client.ForceOwnership)
+		return reconcile.Result{}, a.client.Status().Patch(ctx, wlPatch, client.Apply, client.FieldOwner(ControllerName+"-finish"), client.ForceOwnership)
 	}
 
 	hasReserving, reservingRemote := group.FirstReserving()
 
+	// if the reserving workload is out of sync, remove it
+	if hasReserving {
+		outOfSync := group.local == nil || !equality.Semantic.DeepEqual(group.local.Spec, group.remotes[reservingRemote].Spec)
+		if outOfSync {
+			if err := client.IgnoreNotFound(group.RemoveRemoteObjects(ctx, reservingRemote)); err != nil {
+				log.V(2).Error(err, "Deleting out of sync remote objects", "remote", reservingRemote)
+				return reconcile.Result{}, err
+			}
+			group.remotes[reservingRemote] = nil
+
+			// check if another one exists
+			hasReserving, reservingRemote = group.FirstReserving()
+		}
+	}
+
+	// If there is no reserving and the AC is ready, the connection with the reserving remote might
+	// be lost, keep the workload admitted for keepReadyTimeout and put it back in the queue after that.
+	if !hasReserving && acs.State == kueue.CheckStateReady {
+		if time.Now().Before(acs.LastTransitionTime.Add(a.keepReadyTimeout)) {
+			retryAfter := a.keepReadyTimeout - time.Since(acs.LastTransitionTime.Time)
+			log.V(3).Info("Reserving remote lost, retry", "retryAfter", retryAfter)
+			return reconcile.Result{RequeueAfter: retryAfter}, nil
+		} else {
+			acs.State = kueue.CheckStateRetry
```

Comment:
I'm not sure this is the purpose of the `CheckStateRetry`, could we just keep it Ready?

Reply:
Ok, I have two concerns with that:
1. in the [docs for Retry](https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/apis/kueue/v1beta1/admissioncheck_types.go#L28C2-L30C19
) we say "A workload having at least one check in the state, will be evicted if admitted will not be considered for admission.". However, in the case of workload being temporarily unreachable it is already admitted, so I'm not sure it fits the definition, does it? 

2. In the [docs for Rejected](https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/apis/kueue/v1beta1/admissioncheck_types.go#L35-L37. ) we have "NOTE: The admission behaviour is currently the same as for retry, we can consider marking the workload as "Finished" with a failure description.". Should we update the note then? I'm also concerned about the checks in code which make `Retry` and `Rejected` equivalent, like [here](https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/pkg/workload/admissionchecks.go#L145), so we may need to review all the places if this still makes sense.

Reply:
In other words, I'm not understanding yet was is the difference between `Rejected` and `Retry`, and why `Retry` is preferred. 

After the `WorkerLostTimeout` is exceeded we should requeue the workload. Can you elaborate what is the difference between `Rejected` and `Retry`, and how the choice impacts the scenario.

Reply:
`Rejected` is permanent, the workload is marked as `Finished` and should never go back in the queue

Reply:
Got it, thanks for explaining. Should we update the descriptions to indicate the difference (potentially in this PR, preparatory, or follow up)? This "A workload having at least one check in the state, will be evicted if admitted **will not be considered for admission.**" suggests to me that the workload is not retried also. Similarly, the note suggests "NOTE: The admission behaviour is currently the same as for retry".

Reply:
indeed 

https://github.com/kubernetes-sigs/kueue/blob/fa96518c293473480b0d1f45f642af23e433302e/apis/kueue/v1beta1/admissioncheck_types.go#L35-L38

is outdated, it should be changed in a cleanup  PR

Reply:
#1763